design/service-apis-implementation.md

```diff
@@ -0,0 +1,229 @@
+Project Contour (the organisation) may investigate a Service-APIs based layer 4 solution in the future, but that effort will not be in `projectcontour/contour`.
+
+### Contour Gateway model
+
+Contour considers a GatewayClass and associated set of Gateways to describe a single Envoy deployment, deployed and managed by something outside of itself.
+As such, Contour will error out if it is asked to watch more than one GatewayClass.
+Inside its GatewayClass, Contour will merge Listeners within a single Gateway, or across Gateways in the GatewayClass. (See the detailed design for the exact rules Contour will use for this.)
+
+### Interrelated watches
+
+The Service APIs are a set of interrelated Kubernetes objects, where a change in one object can mean that the scope of objects Contour is interested in will change.
+Because of this, Contour will watch all events associated with GatewayClass, Gateway, HTTPRoute, TLSRoute, and BackendPolicy objects, and filter the objects it takes action on internally, rather than using a filtered watch (as those are costly to set up and tear down).
+
+### Combining Service APIs with other configuration
+
+When it calculates the Envoy configuration using the DAG, Contour layers different types of configuration in order.
+So, currently, Ingress is overwritten by HTTPProxy, so if an Ingress and a HTTPProxy specify the exact same route, the HTTPProxy will win.
+
+Once we have the Service-APIs available as well, we have to choose the order.
+In this design, I suggest having the order be Ingress is overwritten by HTTPProxy, is overwritten by the service-apis.
+I could see reversing HTTPProxy and the Service APIs here, but I think this acknowledges that the Service APIs are really an evolution of the ideas in HTTPProxy, and will probably end up being the community standard.
+
+### Status management
+
+Each object in the Service APIs set has its own status definition. In general, Contour will update the status of any object that comes into scope with its details, ensuring that the `observedGeneration` field is set correctly.
+When objects fall out of scope, any status set by Contour will not be removed.
+It's expected that things that check the status will also check the `observedGeneration` to check if the status information is up-to-date with the current object generation.
+
+### Per-object design notes
+
+#### GatewayClass
+
+Contour will look for entries with `spec.controller` equal to `projectcontour.io/contour` by default.
+Contour can change this value to any `projectcontour.io/<value>` with a config setting (similar to how `ingress.class` works today.)
+Importantly, this means that changing this values requires a Contour restart; this means that the GatewayClass watcher can have this value baked in at runtime.
+
+If Contour finds more than one matching GatewayClass, this is a conflict, and Contour will use the oldest GatewayClass (by creation timestamp).
```

Comment:
it looks like the "oldest object wins" strategy is used in a few places - are there any alternatives? why did you choose this strategy?

Reply:
seems to be some mention of oldest-wins here as well: https://kubernetes-sigs.github.io/service-apis/httproute/#tls (notes section)

Reply:
just found https://kubernetes-sigs.github.io/service-apis/guidelines/#conflicts as well which explicitly mentions this as a principle

Reply:
Yes, this is why I picked this, I think that oldest object wins is about the best we can hope for here.